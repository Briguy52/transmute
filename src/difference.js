import curry from "./curry";

/**
 * Take the difference between one iterable and another iterable.
 * Only the elements present in just subject will remain.
 *
 * @param  {Iterable} toRemove
 * @param  {Iterable} subject
 * @return {Iterable}
 */
function difference(toRemove, subject) {
  if (!toRemove) {
    return subject;
  }
  return subject.filter(value => !toRemove.contains(value));
}

export default curry(difference);
