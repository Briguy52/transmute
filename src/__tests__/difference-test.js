import { List, Set, Seq, Map } from 'immutable';
import difference from '../difference';

describe('transmute/difference', () => {
  it('returns empty iterable when no differences', () => {
    expect(difference(List(['a', 'b']), List(['a', 'b']))).toMatchSnapshot();
  });

  it('returns full iterable when no toRemove given', () => {
    expect(difference(undefined, List(['a', 'b']))).toMatchSnapshot();
  });

  it('returns differences for list', () => {
    expect(difference(List(['c', 'b']), List(['a', 'b']))).toMatchSnapshot();
  });

  it('returns differences for Set', () => {
    expect(difference(Set(['c', 'b']), Set(['a', 'b']))).toMatchSnapshot();
  });

  it('returns differences for Seq', () => {
    expect(difference(Seq(['c', 'b']), Seq(['a', 'b']))).toMatchSnapshot();
  });

  it('returns differences for Map with key from subject', () => {
    expect(
      difference(Map({ c1: 'c', b1: 'b' }), Map({ a2: 'a', b2: 'b' }))
    ).toMatchSnapshot();
  });

  it('return type is same as subject type', () => {
    expect(List.isList(difference(Seq(), List()))).toBe(true);
  });
});
