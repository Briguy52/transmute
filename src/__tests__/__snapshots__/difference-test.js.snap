// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/difference returns differences for Map with key from subject 1`] = `
Immutable.Map {
  a2: "a",
}
`;

exports[`transmute/difference returns differences for Seq 1`] = `
Array [
  "a",
]
`;

exports[`transmute/difference returns differences for Set 1`] = `
Immutable.Set [
  "a",
]
`;

exports[`transmute/difference returns differences for list 1`] = `
Immutable.List [
  "a",
]
`;

exports[`transmute/difference returns empty iterable when no differences 1`] = `
Immutable.List [
]
`;

exports[`transmute/difference returns full iterable when no toRemove given 1`] = `
Immutable.List [
  "a",
  "b",
]
`;
